File: ConvenientFileManager/Documents/NSFileManager+CFMDocuments.m
//
//  NSFileManager+CFMDocuments.m
//  ConvenientFileManager
//
//  Created by William Boles on 10/01/2014.
//  Copyright © 2015 Boles. All rights reserved.
//

#import "NSFileManager+CFMDocuments.h"

#import "NSFileManager+CFMPersistence.h"

@implementation NSFileManager (CFMDocuments)

#pragma mark - Documents

+ (NSString *)cfm_documentsDirectoryPath
{
    NSURL *directoryURL = [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory
                                                                  inDomains:NSUserDomainMask] lastObject];
    
    return [directoryURL path];
}

+ (NSString *)cfm_documentsDirectoryPathForResourceWithPath:(NSString *)relativePath
{
    NSString *cacheDirectory = [NSFileManager cfm_documentsDirectoryPath];
    
    return [cacheDirectory stringByAppendingPathComponent:relativePath];
}

#pragma mark - Saving

+ (BOOL)cfm_saveData:(NSData *)data toDocumentsDirectoryPath:(NSString *)relativePath
{
    BOOL saved = NO;
    
    if (relativePath.length > 0)
    {
        NSString *documentsDirectory = [NSFileManager cfm_documentsDirectoryPath];
        NSString *absolutePath = [documentsDirectory stringByAppendingPathComponent:relativePath];
        
        saved = [NSFileManager cfm_saveData:data
                                     toPath:absolutePath];
    }
    
    return saved;
}

#pragma mark - Retrieval

+ (NSData *)cfm_retrieveDataFromDocumentsDirectoryWithPath:(NSString *)relativePath
{
    NSData *data = nil;
    
    if (relativePath.length > 0)
    {
        NSString *documentsDirectory = [NSFileManager cfm_documentsDirectoryPath];
        NSString *absolutePath = [documentsDirectory stringByAppendingPathComponent:relativePath];
        
        data = [NSFileManager cfm_retrieveDataAtPath:absolutePath];
    }
    
    return data;
}

#pragma mark - Exists

+ (BOOL)cfm_fileExistsInDocumentsDirectory:(NSString *)relativePath
{
    BOOL fileExists = NO;
    
    if (relativePath.length > 0)
    {
        NSString *documentsDirectory = [NSFileManager cfm_documentsDirectoryPath];
        NSString *absolutePath = [documentsDirectory stringByAppendingPathComponent:relativePath];
        
        fileExists = [[NSFileManager defaultManager] fileExistsAtPath:absolutePath];
    }
    
    return fileExists;
}

#pragma mark - Deletion

+ (BOOL)cfm_deleteDataFromDocumentsDirectoryWithPath:(NSString *)relativePath
{
    BOOL deletion = NO;
    
    if (relativePath.length > 0)
    {
        NSString *documentsDirectory = [NSFileManager cfm_documentsDirectoryPath];
        NSString *absolutePath = [documentsDirectory stringByAppendingPathComponent:relativePath];
        
        deletion = [NSFileManager cfm_deleteDataAtPath:absolutePath];
    }
    
    return deletion;
}

@end

File: ConvenientFileManager/Documents/NSFileManager+CFMDocuments.swift
//
//  NSFileManager+Documents.swift
//  ConvenientFileManager
//
//  Created by William Boles on 29/03/2016.
//  Copyright © 2016 Boles. All rights reserved.
//

import Foundation

public extension NSFileManager {

    //MARK: Documents
    
    public class func documentsDirectoryPath() -> String {
        return (NSFileManager.documentsDirectoryURL().path)!
    }
    
    public class func documentsDirectoryURL() -> NSURL {
        return NSFileManager.defaultManager().URLsForDirectory(NSSearchPathDirectory.DocumentDirectory, inDomains: NSSearchPathDomainMask.UserDomainMask).last!
    }
    
    public class func documentsDirectoryPathForResourceWithPath(relativePath: String) -> String {
        let documentsDirectoryPathForResource: String
        
        if relativePath.characters.count > 0 {
            let documentsDirectory = NSFileManager.documentsDirectoryURL()
            let absoluteURL = documentsDirectory.URLByAppendingPathComponent(relativePath)
            
            documentsDirectoryPathForResource = absoluteURL.path!
        } else {
            documentsDirectoryPathForResource = NSFileManager.documentsDirectoryPath()
        }
        
        return documentsDirectoryPathForResource
    }
    
    //MARK: Saving
    
    public class func saveDataToDocumentsDirectory(data: NSData, relativePath: String) -> Bool{
        var saved = false
        
        if relativePath.characters.count > 0 && data.length > 0{
            let documentsDirectory = NSFileManager.documentsDirectoryURL()
            
            if let absolutePath = documentsDirectory.URLByAppendingPathComponent(relativePath).path {
                saved = NSFileManager.saveData(data, absolutePath: absolutePath)
            }
        }
        
        return saved
    }
    
    //MARK: Retrieval
    
    public class func retrieveDataFromDocumentsDirectory(relativePath: String) -> NSData? {
        var data: NSData?
        
        if relativePath.characters.count > 0 {
            let documentsDirectory = NSFileManager.documentsDirectoryURL()
            let absolutePath = documentsDirectory.URLByAppendingPathComponent(relativePath).path!
            
            data = NSFileManager.retrieveDataAtPath(absolutePath)
        }
        
        return data
    }
    
    //MARK: Exists
    
    public class func fileExistsInDocumentsDirectory(relativePath: String) -> Bool {
        var fileExists = false
        
        if relativePath.characters.count > 0 {
            let documentsDirectory = NSFileManager.documentsDirectoryURL()
            let absolutePath = documentsDirectory.URLByAppendingPathComponent(relativePath).path!
            
            fileExists = NSFileManager.fileExistsAtPath(absolutePath)
        }
        
        return fileExists
    }
    
    //MARK: Deletion
    
    public class func deleteDataFromDocumentsDirectory(relativePath: String) -> Bool {
        var deleted = false
        
        if relativePath.characters.count > 0 {
            let documentsDirectory = NSFileManager.documentsDirectoryURL()
            let absolutePath = documentsDirectory.URLByAppendingPathComponent(relativePath).path!
            
            deleted = NSFileManager.deleteDataAtPath(absolutePath)
        }
        
        return deleted
    }
}